### ISSUE TITLE
only 1 version per dist-tag - and sinopia keeps highest version only

### ISSUE THREAD

**Issue by jedi4ever** [rlidwka/sinopia#220]
We use dist-tags(tags) to mark the version used in different environments:

```{
   latest: '1.3.1',
   staging: '1.4.1',
   production: '1.3.1'
}```

sometimes we need lower the version number to rollback.
We expected the npm tag to just overwrite the version, but it seems sinopia only keeps the highest version (semver sort).

https://github.com/rlidwka/sinopia/blob/06bb301babe8d17e1f085d4e356f37f1266228f8/lib/utils.js#L103

With the new `npm dist-tag` command, you can add/remove versions. So I'm wondering if we can just overwrite the version of a tag.

Also it seems 'officially' there an only be 1 version per tag. (dixit @othiym23). So there should never be an array.

**Comment by rlidwka** [comment]
Tags are stored in the array because:

 - Historically, sinopia is designed with an assumption that npm registry might at any point be unreachable
 - `latest` tag is extremely important to npm, and people rarely used any other tags

So if we store just one tag, it might point to a version we can't get. Which will lead to an undesirable installation failure. Here is a scenario when it could happen:

-----

 1. Suppose we have `express@4.8` tagged `latest` and `express@5.0` tagged `beta`.
 2. User types `npm install express`, and `express@4.8` is now in sinopia cache tagged `latest`
 3. `express@4.9` is published and tagged `latest` in npm registry
 4. User types `npm install express@5`. Sinopia polls npm, updates tags, so now `latest` points to `express@4.9` which isn't cached in npm.
5. npm registry goes down
6. User types `npm install express`, and it either fails or installs `express@5` both of which are bad.

But if we use an array, latest tag would point to `["4.8", "4.9"]`, and if 4.9 is unavailable, it'll install 4.8.

-----

I'm not sure if it's relevant anymore, and `semver-sorting` obviously isn't ideal. So maybe I'll revisit this as a part of https://github.com/rlidwka/sinopia/issues/211, it might require large changes anyway.

**Comment by jedi4ever** [comment]
@rlidwka I agree with your reasoning . But the official npm registry doesn't store tag versions as a hash.
If you do `npm info express dist-tags` it returns  `{ latest: '4.12.3', rc: '4.0.0-rc4' }` which is not an array.

My preference would be for sinopia to mimic official behavior. Or at least have a flag to configure it.
If interested I can provide this patch

**Comment by jakepusateri** [comment]
@rlidwka In your scenario, I imagine the right thing to do would be to grab and cache all versions referenced in dist-tags and have them available upon update. If you want to avoid the situation described, there shouldn't be a state where the metadata doesn't match what's available.

**Comment by rlidwka** [comment]
@jakepusateri , hmm... that's actually a nice compromise between downloading all tarballs for the package, and downloading only needed ones. Thanks for the idea.

It doesn't guarantee consistency though. I mean, if sinopia waits to download all tarballs before responding to a regular `/package` request, it'll be an unacceptably long wait. But if it doesn't, npm registry might go down, and you'll end up with metadata without actual tarball with it.

And pre-fetching multiple versions (one per tag) of some kind of 40Mb unicode bundle is a pain. Not sure if it's worth it.
